### PR TITLE
[SPARK-35849][PYTHON] Make `astype` method data-type-based for DecimalOps

### PR DESCRIPTION
**Pull request by Yikun**
### What changes were proposed in this pull request?
Make DecimalOps astype data-type-based.

See more in:
https://github.com/apache/spark/pull/32821#issuecomment-861119905

### Why are the changes needed?
Make DecimalOps astype data-type-based.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing test NumOpsTest.test_astype in pyspark/pandas/tests/data_type_ops/test_num_ops.py

### CODE REVIEW

**Comment by Yikun** [comment]
cc @xinrong-databricks @ueshin

File: python/pyspark/pandas/data_type_ops/num_ops.py [hunk 1]
```diff
@@ -416,6 +409,29 @@ def pretty_name(self) -> str:
...
+            if isinstance(dtype, extension_dtypes):
+                scol = index_ops.spark.column.cast(spark_type)
+            else:
+                scol = F.when(index_ops.spark.column.isNull(), F.lit(False)).otherwise(
+                    index_ops.spark.column.cast(spark_type)
+                )
+            return index_ops._with_new_scol(
+                scol.alias(index_ops._internal.data_spark_column_names[0]),
+                field=InternalField(dtype=dtype),
+            )
```
Comment (ueshin) [inline python/pyspark/pandas/data_type_ops/num_ops.py:429]: I guess we can reuse `_as_bool_type` in `base.py` file?

**Review by ueshin — COMMENTED** [review]
Good catch!
Otherwise, LGTM, pending tests.

File: python/pyspark/pandas/data_type_ops/num_ops.py (hunk 1 above)
Comment (Yikun) [reply python/pyspark/pandas/data_type_ops/num_ops.py]: yep

**Review by ueshin — APPROVED** [review]

**Comment by ueshin** [comment]
Thanks! merging to master.

**Comment by xinrong-databricks** [comment]
Thanks @Yikun !

**Pull request closed without merging**